### PR TITLE
Avoid double-free on unsuccessful getting PRNG seeding

### LINKED ISSUE #16331: A possible divide by zero bug in p12_key.c
In the file `p12_key.c`, the function `PKCS12_key_gen_uni` has the following [code](https://github.com/openssl/openssl/blob/bc8c36272067f8443f875164831ce3a5a739df3f/crypto/pkcs12/p12_key.c#L148-#L149):
```
u = EVP_MD_size(md_type);
if (u < 0 || v <= 0)
    goto err;
...
for (j = 0; j < v; j++)
    B[j] = Ai[j % u];
```

The variable `u` is used as a divisor. Since the code explicitly checks whether `u` is negative, I think it should also avoid the case when `u=0` to prevent divide by zero bugs.

### LINKED ISSUE #16631: getrandom EAGAIN is not processed correctly
The command 
`LD_LIBRARY_PATH=. strace -e inject=getrandom:error=EAGAIN apps/openssl rand -out ./testrand 1024`

causes a double-free with the following stack trace:
```
#0  0x00007f6d1c1aa9e5 in raise () from /lib64/libc.so.6
#1  0x00007f6d1c1938a4 in abort () from /lib64/libc.so.6
#2  0x00007f6d1c1ed007 in __libc_message () from /lib64/libc.so.6
#3  0x00007f6d1c1f4cdc in malloc_printerr () from /lib64/libc.so.6
#4  0x00007f6d1c1f67ac in _int_free () from /lib64/libc.so.6
#5  0x00007f6d1c698820 in cleanup_entropy (outlen=0, out=0xa30490 "\020\017\243", drbg=0x9d32f0) at providers/implementations/rands/drbg.c:259
#6  cleanup_entropy (outlen=0, out=0xa30490 "\020\017\243", drbg=0x9d32f0) at providers/implementations/rands/drbg.c:248
#7  ossl_prov_drbg_instantiate (drbg=0x9d32f0, strength=<optimized out>, prediction_resistance=<optimized out>, pers=<optimized out>, perslen=<optimized out>) at providers/implementations/rands/drbg.c:473
#8  0x00007f6d1c5640a4 in evp_rand_instantiate_locked (params=0x7ffd52b3bcb0, pstr_len=<optimized out>, pstr=0x0, prediction_resistance=0, strength=0, ctx=0x9b9da0) at crypto/evp/evp_rand.c:505
#9  EVP_RAND_instantiate (ctx=ctx@entry=0x9b9da0, strength=strength@entry=0, prediction_resistance=prediction_resistance@entry=0, pstr=pstr@entry=0x0, pstr_len=pstr_len@entry=0, params=params@entry=0x7ffd52b3bcb0)
    at crypto/evp/evp_rand.c:518
#10 0x00007f6d1c5c5d60 in rand_new_drbg (libctx=libctx@entry=0x0, parent=<optimized out>, reseed_interval=<optimized out>, reseed_interval@entry=256, reseed_time_interval=<optimized out>, reseed_time_interval@entry=3600, 
    use_df=<optimized out>, use_df@entry=1) at crypto/rand/rand_lib.c:599
#11 0x00007f6d1c5c685e in RAND_get0_primary (ctx=ctx@entry=0x0) at crypto/rand/rand_lib.c:646
#12 0x00007f6d1c5c6bd8 in RAND_get0_public (ctx=ctx@entry=0x0) at crypto/rand/rand_lib.c:677
#13 0x00007f6d1c5c6c80 in RAND_bytes_ex (ctx=0x0, buf=0x7ffd52b3bee0 "P\003", num=1024, strength=0) at crypto/rand/rand_lib.c:362
#14 0x000000000044d634 in rand_main (argc=<optimized out>, argv=<optimized out>) at apps/rand.c:123
#15 0x0000000000445078 in do_cmd (prog=prog@entry=0x9b0850, argc=argc@entry=4, argv=argv@entry=0x7ffd52b3d0d0) at apps/openssl.c:414
#16 0x0000000000422507 in main (argc=4, argv=0x7ffd52b3d0d0) at apps/openssl.c:295
```

### PR DESCRIPTION
**Pull request by beldmit**
Fixes #16631

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: providers/implementations/rands/seed_src.c [hunk 1]
```diff
@@ -206,6 +206,7 @@ static size_t seed_get_seed(void *vseed, unsigned char **pout,
                           adin, adin_len) != 0)
         return bytes_needed;
     OPENSSL_secure_clear_free(p, bytes_needed);
+    *pout = NULL;
```
Comment (mattcaswell) [inline providers/implementations/rands/seed_src.c:209]: Would it be better to only set `pout` in the case the call to `seed_src_generate` succeeds?

Comment (beldmit) [reply providers/implementations/rands/seed_src.c:209]: As you wish. I don't see any significant difference, though your approach looks a bit more safe.

Comment (beldmit) [reply providers/implementations/rands/seed_src.c]: Done, thanks!

**Comment by beldmit** [comment]
Looks like other RNGs are not affected.

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
I think it would also be worthwhile fixing the caller to not free this pointer on error.

I've created #16640 to do that.  I think the two should stand independently.

**Review by t8m — APPROVED** [review]

**Pull request closed without merging**